FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1323 @@
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackMember1.StackDataNode, StackMember2.StackDataNode);
+          }
+        }
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+      case OP_F80STACKTEST: {
+        const auto* Op = IROp->C<IROp_F80StackTest>();
+        auto offset = Op->SrcStack;
+        const auto& [Valid, StackMember] = StackData.top(offset);
+
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        Ref CmpNode {};
+        Ref ZeroConst = IREmit->_VCastFromGPR(ReducedPrecisionMode ? 8 : 16, 8, GetConstant(0));
+
+        if (SlowPath) {
+          auto* StackValue = LoadStackValueAtOffset_Slow(offset);
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackValue, ZeroConst);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackValue, ZeroConst);
+          }
+        } else {
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackMember.StackDataNode, ZeroConst);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackMember.StackDataNode, ZeroConst);
+          }
+        }
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+
+      case OP_F80CMPVALUE: {
+        const auto* Op = IROp->C<IROp_F80CmpValue>();
+        const auto& Value = CurrentIR.GetNode(Op->X80Src);
+        const auto& [Valid, StackMember] = StackData.top();
+
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        Ref CmpNode {};
+        if (SlowPath) {
+          auto* StackValue = LoadStackValueAtTop_Slow();
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackValue, Value);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackValue, Value);
+          }
+        } else {
+          if (ReducedPrecisionMode) {
+            CmpNode = IREmit->_FCmp(8, StackMember.StackDataNode, Value);
+          } else {
+            CmpNode = IREmit->_F80Cmp(StackMember.StackDataNode, Value);
+          }
+        }
+
+        IREmit->ReplaceUsesWithAfter(CodeNode, CmpNode, CodeNode);
+        IREmit->Remove(CodeNode);
+        break;
+      }
+
+      case OP_F80XTRACTSTACK: {
+        const auto& [Valid, StackMember] = StackData.top();
+
+        MigrateToSlowPathIf(Valid != StackSlot::VALID);
+
+        if (SlowPath) {
+          Ref st0 = LoadStackValueAtTop_Slow();
+          Ref Exp {};
+          Ref Sig {};
+          if (ReducedPrecisionMode) {
+            std::tie(Exp, Sig) = SplitF64SigExp(st0);
+          } else {
+            Exp = IREmit->_F80XTRACT_EXP(st0);
+            Sig = IREmit->_F80XTRACT_SIG(st0);
+          }
+          // Write exp to top, update top for a push and set sig at new top.
+          StoreStackValueAtTop_Slow(Exp, false);
+          UpdateTopForPush_Slow();
```

Comment:
Likewise probably want a StackPush helper for something like:

```
if (Slow) {
          UpdateTopForPush_Slow();
          StoreStackValueAtTop_Slow(Ref);
} else {
         StackData.push(StackMemberInfo {.SourceDataSize = StackMember.SourceDataSize,
                                          .SourceDataNode = nullptr,
                                          .StackDataNode = Ref,
                                          .InterpretAsFloat = StackMember.InterpretAsFloat});
```

and then with the other helpers the branching here all goes away